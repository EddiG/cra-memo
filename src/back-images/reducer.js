// @flow
import { combineReducers } from 'redux';
import createReducer from '../utils/reducers';
import yellowImage from './yellow-bg.jpg';

type Image = {
  id: string,
  src: string,
};

type Images = {
  [string]: Image,
};

const initialStateById: Images = {
  backImage1: {
    id: 'backImage1',
    // src: 'https://dummyimage.com/200x200/957fae/fff.png&text=+',
    src: yellowImage,
  },
};

const initialStateAllIds: Array<string> = ['backImage1'];

const byId = createReducer(initialStateById, {});

const allIds = createReducer(initialStateAllIds, {});

const backImagesReducer = combineReducers({
  byId,
  allIds,
});

export default backImagesReducer;
