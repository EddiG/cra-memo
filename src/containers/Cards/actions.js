// @flow
import * as constants from './constants';
import type { Card } from './types';

export const updateCards = (cards: Array<Card>) => ({
  type: constants.UPDATE_CARDS,
  payload: {
    cards,
  },
});

export const closeCards = () => ({
  type: constants.CLOSE_CARDS,
  payload: {},
});

export const triggerCard = (cardId: String) => ({
  type: constants.TRIGGER_CARD,
  payload: {
    id: cardId,
  },
});

export const triggerCardAsync = (cardId: String) => ({
  type: constants.TRIGGER_CARD_ASYNC,
  payload: {
    id: cardId,
  },
});

export const validateCards = () => ({
  type: constants.VALIDATE_CARDS,
  payload: {},
});
