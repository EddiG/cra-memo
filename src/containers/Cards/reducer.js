// @flow
import { combineReducers } from 'redux';
import assign from 'lodash/assign';
import filter from 'lodash/filter';
import reduce from 'lodash/reduce';
import createReducer from '../../utils/reducers';
import * as constants from './constants';
import type { Card } from './types';

type Cards = {
  [string]: Card,
};

type Ids = Array<string>;

const updateCardsById = (state: Cards, action): Cards => {
  const { cards } = action.payload;
  return cards.reduce((acc, card) => ({ ...acc, ...{ [card.id]: card } }), {});
};

const updateCardsAllIds = (state: Ids, action): Ids => {
  const { cards } = action.payload;
  return cards.map(card => card.id);
};

const validateCards = (state: Cards): Cards => {
  const openedCards = filter(state, card => !card.isGuessed && card.isOpened);

  const validatedCards = openedCards.map(card => {
    const pairCard = openedCards.filter(c => c.pairCardId === card.id);
    if (pairCard.length) {
      return { ...card, isGuessed: true };
    }
    return { ...card, isOpened: false };
  });

  return assign(
    {},
    state,
    validatedCards.reduce((acc, card) => ({ ...acc, [card.id]: card }), {}),
  );
};

const closeCards = (state: Cards): Cards =>
  reduce(
    state,
    (acc, card) => ({
      ...acc,
      [card.id]: { ...card, isOpened: false, isGuessed: false },
    }),
    {},
  );

const triggerCard = (state: Cards, action): Cards => {
  const { id } = action.payload;
  const card = state[id];
  return assign({}, state, {
    [card.id]: { ...card, isOpened: card.isGuessed ? true : !card.isOpened },
  });
};

const byId = createReducer(({}: Cards), {
  [constants.UPDATE_CARDS]: updateCardsById,
  [constants.CLOSE_CARDS]: closeCards,
  [constants.TRIGGER_CARD]: triggerCard,
  [constants.VALIDATE_CARDS]: validateCards,
});

const allIds = createReducer(([]: Ids), {
  [constants.UPDATE_CARDS]: updateCardsAllIds,
});

export default combineReducers({
  byId,
  allIds,
});
