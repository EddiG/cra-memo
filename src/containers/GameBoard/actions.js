import * as constants from './constants';

export const generateCards = () => ({
  type: constants.GENERATE_CARDS,
  payload: {},
});

export const refreshCards = () => ({
  type: constants.REFRESH_CARDS,
  payload: {},
});
