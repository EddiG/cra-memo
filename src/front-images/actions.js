import * as constants from './constants';

export const preloadImagesRequire = () => ({
  type: constants.PRELOAD_FRONT_IMAGES_REQUIRE,
  payload: {},
});

export const preloadImagesComplete = () => ({
  type: constants.PRELOAD_FRONT_IMAGES_COMPLETE,
  payload: {},
});
