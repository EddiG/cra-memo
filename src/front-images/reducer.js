// @flow
import { combineReducers } from 'redux';
import createReducer from '../utils/reducers';
import pooh from './winnie/pooh.png';
import pigglet from './winnie/pigglet.png';
import tiger from './winnie/tiger.png';
import ia from './winnie/ia.png';
import owl from './winnie/owl.png';
import rabbit from './winnie/rabbit.png';
import kenga from './winnie/kenga.png';

type Image = {
  id: string,
  src: string,
};

type Images = {
  [string]: Image,
};

const initialStateById: Images = {
  frontImage1: {
    id: 'frontImage1',
    src: pooh,
  },
  frontImage2: {
    id: 'frontImage2',
    src: pigglet,
  },
  frontImage3: {
    id: 'frontImage3',
    src: tiger,
  },
  frontImage4: {
    id: 'frontImage4',
    src: ia,
  },
  frontImage5: {
    id: 'frontImage5',
    src: owl,
  },
  frontImage6: {
    id: 'frontImage6',
    src: kenga,
  },
  frontImage7: {
    id: 'frontImage7',
    src: rabbit,
  },
};

const initialStateAllIds: Array<string> = [
  'frontImage1',
  'frontImage2',
  'frontImage3',
  'frontImage4',
  'frontImage5',
  'frontImage6',
  'frontImage7',
];

const byId = createReducer(initialStateById, {});

const allIds = createReducer(initialStateAllIds, {});

const frontImagesReducer = combineReducers({
  byId,
  allIds,
});

export default frontImagesReducer;
