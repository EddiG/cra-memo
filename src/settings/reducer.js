// @flow
import createReducer from '../utils/reducers';

type Settings = {
  cardsQuantity: number,
};

const initialState: Settings = {
  cardsQuantity: 12,
};

const settingsReducer = createReducer(initialState, {});

export default settingsReducer;
