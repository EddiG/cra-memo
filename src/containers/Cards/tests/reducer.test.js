import reducer from '../reducer';
import * as actions from '../actions';

describe('cards reducer', () => {
  it('should handle UPDATE_CARDS', () => {
    const cards = [
      {
        id: 'card1',
        isOpened: false,
        isGuessed: false,
        frontImageId: 'frontImage1',
        backImageId: 'backImage1',
      },
      {
        id: 'card2',
        isOpened: false,
        isGuessed: false,
        frontImageId: 'frontImage2',
        backImageId: 'backImage2',
      },
    ];
    const updateAction = actions.updateCards(cards);
    expect(reducer({}, updateAction)).toMatchSnapshot();
  });
  it('should handle CLOSE_CARDS', () => {
    const state = {
      byId: {
        card10: {
          id: 'card10',
          isOpened: true,
          isGuessed: false,
          frontImageId: 'frontImage1',
          backImageId: 'backImage1',
        },
        card20: {
          id: 'card20',
          isOpened: false,
          isGuessed: false,
          frontImageId: 'frontImage1',
          backImageId: 'backImage1',
        },
      },
      allIds: ['card10', 'card20'],
    };
    const closeCardsAction = actions.closeCards();
    expect(reducer(state, closeCardsAction)).toMatchSnapshot();
  });
  it('should handle TRIGGER_CARD', () => {
    const state = {
      byId: {
        card10: {
          id: 'card10',
          isOpened: false,
          isGuessed: false,
          frontImageId: 'frontImage1',
          backImageId: 'backImage1',
        },
        card20: {
          id: 'card20',
          isOpened: false,
          isGuessed: false,
          frontImageId: 'frontImage1',
          backImageId: 'backImage1',
        },
      },
      allIds: ['card10', 'card20'],
    };
    const triggerCard = actions.triggerCard('card10');
    expect(reducer(state, triggerCard)).toMatchSnapshot();
  });
  it('should handle VALIDATE_CARDS when the not paired cards was opened', () => {
    const state = {
      byId: {
        card10: {
          id: 'card10',
          pairCardId: 'card20',
          isOpened: true,
          isGuessed: false,
          frontImageId: 'frontImage1',
          backImageId: 'backImage1',
        },
        card20: {
          id: 'card20',
          pairCardId: 'card10',
          isOpened: false,
          isGuessed: false,
          frontImageId: 'frontImage1',
          backImageId: 'backImage1',
        },
        card30: {
          id: 'card30',
          pairCardId: 'card40',
          isOpened: true,
          isGuessed: false,
          frontImageId: 'frontImage11',
          backImageId: 'backImage1',
        },
        card40: {
          id: 'card40',
          pairCardId: 'card30',
          isOpened: false,
          isGuessed: false,
          frontImageId: 'frontImage11',
          backImageId: 'backImage1',
        },
        card50: {
          id: 'card50',
          pairCardId: 'card60',
          isOpened: true,
          isGuessed: false,
          frontImageId: 'frontImage21',
          backImageId: 'backImage1',
        },
        card60: {
          id: 'card60',
          pairCardId: 'card50',
          isOpened: false,
          isGuessed: false,
          frontImageId: 'frontImage21',
          backImageId: 'backImage1',
        },
      },
      allIds: ['card10', 'card20', 'card30', 'card40', 'card50', 'card60'],
    };
    const validateCards = actions.validateCards();
    expect(reducer(state, validateCards)).toMatchSnapshot();
  });
  it('should handle VALIDATE_CARDS when the paired cards was opened', () => {
    const state = {
      byId: {
        card10: {
          id: 'card10',
          pairCardId: 'card20',
          isOpened: true,
          isGuessed: false,
          frontImageId: 'frontImage1',
          backImageId: 'backImage1',
        },
        card20: {
          id: 'card20',
          pairCardId: 'card10',
          isOpened: true,
          isGuessed: false,
          frontImageId: 'frontImage1',
          backImageId: 'backImage1',
        },
        card30: {
          id: 'card30',
          pairCardId: 'card40',
          isOpened: false,
          isGuessed: false,
          frontImageId: 'frontImage11',
          backImageId: 'backImage1',
        },
        card40: {
          id: 'card40',
          pairCardId: 'card30',
          isOpened: false,
          isGuessed: false,
          frontImageId: 'frontImage11',
          backImageId: 'backImage1',
        },
      },
      allIds: ['card10', 'card20', 'card30', 'card40'],
    };
    const validateCards = actions.validateCards();
    expect(reducer(state, validateCards)).toMatchSnapshot();
  });
});
