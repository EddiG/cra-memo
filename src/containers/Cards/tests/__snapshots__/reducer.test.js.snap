// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cards reducer should handle CLOSE_CARDS 1`] = `
Object {
  "allIds": Array [
    "card10",
    "card20",
  ],
  "byId": Object {
    "card10": Object {
      "backImageId": "backImage1",
      "frontImageId": "frontImage1",
      "id": "card10",
      "isGuessed": false,
      "isOpened": false,
    },
    "card20": Object {
      "backImageId": "backImage1",
      "frontImageId": "frontImage1",
      "id": "card20",
      "isGuessed": false,
      "isOpened": false,
    },
  },
}
`;

exports[`cards reducer should handle TRIGGER_CARD 1`] = `
Object {
  "allIds": Array [
    "card10",
    "card20",
  ],
  "byId": Object {
    "card10": Object {
      "backImageId": "backImage1",
      "frontImageId": "frontImage1",
      "id": "card10",
      "isGuessed": false,
      "isOpened": true,
    },
    "card20": Object {
      "backImageId": "backImage1",
      "frontImageId": "frontImage1",
      "id": "card20",
      "isGuessed": false,
      "isOpened": false,
    },
  },
}
`;

exports[`cards reducer should handle UPDATE_CARDS 1`] = `
Object {
  "allIds": Array [
    "card1",
    "card2",
  ],
  "byId": Object {
    "card1": Object {
      "backImageId": "backImage1",
      "frontImageId": "frontImage1",
      "id": "card1",
      "isGuessed": false,
      "isOpened": false,
    },
    "card2": Object {
      "backImageId": "backImage2",
      "frontImageId": "frontImage2",
      "id": "card2",
      "isGuessed": false,
      "isOpened": false,
    },
  },
}
`;

exports[`cards reducer should handle VALIDATE_CARDS when the not paired cards was opened 1`] = `
Object {
  "allIds": Array [
    "card10",
    "card20",
    "card30",
    "card40",
    "card50",
    "card60",
  ],
  "byId": Object {
    "card10": Object {
      "backImageId": "backImage1",
      "frontImageId": "frontImage1",
      "id": "card10",
      "isGuessed": false,
      "isOpened": false,
      "pairCardId": "card20",
    },
    "card20": Object {
      "backImageId": "backImage1",
      "frontImageId": "frontImage1",
      "id": "card20",
      "isGuessed": false,
      "isOpened": false,
      "pairCardId": "card10",
    },
    "card30": Object {
      "backImageId": "backImage1",
      "frontImageId": "frontImage11",
      "id": "card30",
      "isGuessed": false,
      "isOpened": false,
      "pairCardId": "card40",
    },
    "card40": Object {
      "backImageId": "backImage1",
      "frontImageId": "frontImage11",
      "id": "card40",
      "isGuessed": false,
      "isOpened": false,
      "pairCardId": "card30",
    },
    "card50": Object {
      "backImageId": "backImage1",
      "frontImageId": "frontImage21",
      "id": "card50",
      "isGuessed": false,
      "isOpened": false,
      "pairCardId": "card60",
    },
    "card60": Object {
      "backImageId": "backImage1",
      "frontImageId": "frontImage21",
      "id": "card60",
      "isGuessed": false,
      "isOpened": false,
      "pairCardId": "card50",
    },
  },
}
`;

exports[`cards reducer should handle VALIDATE_CARDS when the paired cards was opened 1`] = `
Object {
  "allIds": Array [
    "card10",
    "card20",
    "card30",
    "card40",
  ],
  "byId": Object {
    "card10": Object {
      "backImageId": "backImage1",
      "frontImageId": "frontImage1",
      "id": "card10",
      "isGuessed": true,
      "isOpened": true,
      "pairCardId": "card20",
    },
    "card20": Object {
      "backImageId": "backImage1",
      "frontImageId": "frontImage1",
      "id": "card20",
      "isGuessed": true,
      "isOpened": true,
      "pairCardId": "card10",
    },
    "card30": Object {
      "backImageId": "backImage1",
      "frontImageId": "frontImage11",
      "id": "card30",
      "isGuessed": false,
      "isOpened": false,
      "pairCardId": "card40",
    },
    "card40": Object {
      "backImageId": "backImage1",
      "frontImageId": "frontImage11",
      "id": "card40",
      "isGuessed": false,
      "isOpened": false,
      "pairCardId": "card30",
    },
  },
}
`;
